### ISSUE TITLE
bizarre constant mangling in 5.8.0

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5786]
Migrated from [rt.perl.org#15654](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=15654) (status was 'resolved')

Searchable as RT15654$

**Comment by p5pRT** [comment]
### From troc@netrus.net

#### Created by troc@eyrie.homenet



Tests for the POE 0\.22 distribution \(on the CPAN\) have been working
fine with perl 5\.6\.1 for several weeks\.  They pass for perl 5\.8\.0\, but
they generate a disturbing warning in the process&#8203;:

2\) eyrie&#8203;:\~/perl/poe% /usr/local/perl\-580/bin/perl t/08\_errors\.t
1\.\.59
ok 1
\.\.\.\.\.
ok 17
Argument "\,\-\-\-\-\- SELECT BITS IN \-\-\-\-\-\\n" isn't numeric in array \\
&nbsp;        element at /home/troc/perl/poe/POE/Kernel\.pm line 1798\.
ok 18
\.\.\.\.\.\.

This behavior has been verified with perl 5\.8\.0 on Linux\, MacOS X\, and
FreeBSD\.  The test program is t/08\_errors\.t\, which can be run from the
distribution's tarball directory as&#8203;:

&nbsp;  perl580 \-I\. t/08\_errors\.t

The code that's failing is actually a few lines further than the
warning says\.  I've verified its position while looking through the
perl \-Dt output \(excerpt below\)\.

POE&#8203;::Kernel
1807&#8203;:    elsif \($time >= $kr\_events\[\-1\]\->\[ST\_TIME\]\) \{
1808&#8203;:      push @&#8203;kr\_events\, $event\_to\_enqueue;
1809&#8203;:    \}

It turns out that ST\_TIME is being overwritten with a constant value
from some unrelated code\.  In this case\, from POE&#8203;::Kernel&#8203;::Select\, at
around line 263\.  The constant string is not used anywhere else in the
program\.

POE&#8203;::Kernel
 252&#8203;: sub ST\_TIME       \(\) \{ 5 \}  \#   $event\_due\_time\,

POE&#8203;::Kernel&#8203;::Select
 262&#8203;:  if \(TRACE\_SELECT\) \{
 263&#8203;:    warn "\,\-\-\-\-\- SELECT BITS IN \-\-\-\-\-\\n";
 264&#8203;:    warn "| READ    : "\, unpack\('b\*'\, $kr\_vectors\[VEC\_RD\]\)\, "\\n";
 265&#8203;:    warn "| WRITE   : "\, unpack\('b\*'\, $kr\_vectors\[VEC\_WR\]\)\, "\\n";
 266&#8203;:    warn "| EXPEDITE&#8203;: "\, unpack\('b\*'\, $kr\_vectors\[VEC\_EX\]\)\, "\\n";
 267&#8203;:    warn "\`\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\\n";
 268&#8203;:  \}

This bug seems closely related to the one reported in
http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2001\-07/msg01138\.html

Juicy \-Dt output for 5\.8\.0&#8203;:

\(POE/Kernel\.pm&#8203;:1798\)    padav\(@&#8203;kr\_events\)
\(POE/Kernel\.pm&#8203;:1798\)    not
\(POE/Kernel\.pm&#8203;:1798\)    cond\_expr
\(POE/Kernel\.pm&#8203;:1798\)    padsv\($time\)
\(POE/Kernel\.pm&#8203;:1798\)    padav\(@&#8203;kr\_events\)
\(POE/Kernel\.pm&#8203;:1798\)    const\(PVNV\(\-1\)\)
\(POE/Kernel\.pm&#8203;:1798\)    aelem
\(POE/Kernel\.pm&#8203;:1798\)    rv2av
\(POE/Kernel\.pm&#8203;:1798\)    const\(PVIV\("\,\-\-\-\-\- SELECT BITS IN \-\-\-\-\-\\12"\\0\)\)
\(POE/Kernel\.pm&#8203;:1798\)    aelem
Argument "\,\-\-\-\-\- SELECT BITS IN \-\-\-\-\-\\n" isn't numeric in array \\
&nbsp;        element at POE/Kernel\.pm line 1798\.
\(POE/Kernel\.pm&#8203;:1798\)    ge
\(POE/Kernel\.pm&#8203;:1798\)    cond\_expr
\(POE/Kernel\.pm&#8203;:1798\)    pushmark
\(POE/Kernel\.pm&#8203;:1798\)    padav\(@&#8203;kr\_events\)
\(POE/Kernel\.pm&#8203;:1798\)    padsv\($event\_to\_enqueue\)
\(POE/Kernel\.pm&#8203;:1798\)    push
\(POE/Kernel\.pm&#8203;:1798\)    nextstate

And corresponding output from 5\.6\.1&#8203;:

\(POE/Kernel\.pm&#8203;:1798\)    padav\[38\]
\(POE/Kernel\.pm&#8203;:1798\)    not
\(POE/Kernel\.pm&#8203;:1798\)    cond\_expr
\(POE/Kernel\.pm&#8203;:1798\)    padsv\[7\]
\(POE/Kernel\.pm&#8203;:1798\)    padav\[39\]
\(POE/Kernel\.pm&#8203;:1798\)    const\(IV\(\-1\)\)
\(POE/Kernel\.pm&#8203;:1798\)    aelem
\(POE/Kernel\.pm&#8203;:1798\)    rv2av
\(POE/Kernel\.pm&#8203;:1798\)    const\(IV\(5\)\)
\(POE/Kernel\.pm&#8203;:1798\)    aelem
\(POE/Kernel\.pm&#8203;:1798\)    ge
\(POE/Kernel\.pm&#8203;:1798\)    cond\_expr
\(POE/Kernel\.pm&#8203;:1798\)    null
\(POE/Kernel\.pm&#8203;:1798\)    pushmark
\(POE/Kernel\.pm&#8203;:1798\)    padav\[39\]
\(POE/Kernel\.pm&#8203;:1798\)    padsv\[34\]
\(POE/Kernel\.pm&#8203;:1798\)    push
\(POE/Kernel\.pm&#8203;:1798\)    scope
\(POE/Kernel\.pm&#8203;:1798\)    null
\(POE/Kernel\.pm&#8203;:1798\)    null
\(POE/Kernel\.pm&#8203;:1798\)    nextstate

\-\- Rocco Caputo / troc@&#8203;pobox\.com / poe\.perl\.org / poe\.sf\.net

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl v5.6.1:

Configured by troc at Mon Apr 16 12:13:07 EDT 2001.

Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
  Platform:
    osname=freebsd, osvers=4.3-rc, archname=i386-freebsd
    uname='freebsd eyrie.homenet 4.3-rc freebsd 4.3-rc #0: wed apr 11 01:15:54 edt 2001 troc@eyrie.homenet:usrobjusrsrcsysrc20010411 i386 '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include',
    optimize='-O2 -g',
    cppflags='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.3 [FreeBSD] 20010315 (release)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags ='-Wl,-E  -L/usr/local/lib'
    libpth=/usr/lib /usr/local/lib
    libs=-lm -lc -lcrypt -liconv -lutil
    perllibs=-lm -lc -lcrypt -liconv -lutil
    libc=, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-DPIC -fpic', lddlflags='-shared  -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.1:
    /home/troc/perl/poe
    /usr/local/perl-561-single/lib/i386-freebsd
    /usr/local/perl-561-single/lib
    /usr/local/perl-561-single/site-lib/i386-freebsd
    /usr/local/perl-561-single/site-lib
    /usr/local/perl-561-single/site-lib
    .


Environment for perl v5.6.1:
    HOME=/home/troc
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/usr/games:/usr/X11R6/bin:/home/troc/bin:/usr/local/xt
    PERL5LIB=/home/troc/perl/poe
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From troc@netrus.net

Whoops\!  perlbug on my system is still from 5\.6\.1\.  Here's \-V from
perl 5\.8\.0\.

Summary of my perl5 \(revision 5\.0 version 8 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=freebsd\, osvers=4\.6\-stable\, archname=i386\-freebsd\-thread\-multi
&nbsp;    uname='freebsd eyrie\.homenet 4\.6\-stable freebsd 4\.6\-stable \#0&#8203;: wed jul 10 17&#8203;:26&#8203;:59 edt 2002 troc@&#8203;eyrie\.homenet&#8203;:usrobjusrsrcsysrc20020709 i386 '
&nbsp;    config\_args='\-A define&#8203;:optimize=\-O2 \-g \-pipe \-A define&#8203;:prefix=/usr/local/perl\-580 \-A define&#8203;:privlib=/usr/local/perl\-580/lib \-A define&#8203;:sitelib=/usr/local/perl\-580/site\-lib \-A define&#8203;:installusrbinperl=n \-A define&#8203;:cf\_email=troc@&#8203;pobox\.com \-A define&#8203;:versiononly=n \-D usereentrant \-D usethreads \-D useithreads \-D MULTIPLICITY \-D DEBUGGING \-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-DHAS\_FPSETMASK \-DHAS\_FLOATINGPOINT\_H \-DDEBUGGING \-fno\-strict\-aliasing \-I/usr/local/include'\,
&nbsp;    optimize='\-O2 \-g \-pipe'\,
&nbsp;    cppflags='\-DHAS\_FPSETMASK \-DHAS\_FLOATINGPOINT\_H \-DDEBUGGING \-fno\-strict\-aliasing \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='2\.95\.4 20020320 \[FreeBSD\]'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags ='\-pthread \-Wl\,\-E  \-L/usr/local/lib'
&nbsp;    libpth=/usr/lib /usr/local/lib
&nbsp;    libs=\-lgdbm \-lm \-lc\_r \-lcrypt \-lutil
&nbsp;    perllibs=\-lm \-lc\_r \-lcrypt \-lutil
&nbsp;    libc=\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags='\-DPIC \-fpic'\, lddlflags='\-shared  \-L/usr/local/lib'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: DEBUGGING MULTIPLICITY USE\_ITHREADS USE\_LARGE\_FILES PERL\_IMPLICIT\_CONTEXT
&nbsp;  Built under freebsd
&nbsp;  Compiled at Jul 23 2002 19&#8203;:01&#8203;:19
&nbsp;  %ENV&#8203;:
&nbsp;    PERL5LIB="/home/troc/perl/poe"
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /home/troc/perl/poe
&nbsp;    /usr/local/perl\-580/lib/i386\-freebsd\-thread\-multi
&nbsp;    /usr/local/perl\-580/lib
&nbsp;    /usr/local/perl\-580/site\-lib/i386\-freebsd\-thread\-multi
&nbsp;    /usr/local/perl\-580/site\-lib
&nbsp;    /usr/local/perl\-580/site\-lib
&nbsp;    \.

**Comment by p5pRT** [comment]
### From @hvds

Rocco Caputo \(via RT\) \<perlbug@&#8203;perl\.org> wrote&#8203;:
:Tests for the POE 0\.22 distribution \(on the CPAN\) have been working
:fine with perl 5\.6\.1 for several weeks\.  They pass for perl 5\.8\.0\, but
:they generate a disturbing warning in the process&#8203;:
:
:2\) eyrie&#8203;:\~/perl/poe% /usr/local/perl\-580/bin/perl t/08\_errors\.t
:1\.\.59
:ok 1
:\.\.\.\.\.\.\.
:ok 17
:Argument "\,\-\-\-\-\- SELECT BITS IN \-\-\-\-\-\\n" isn't numeric in array \\
:        element at /home/troc/perl/poe/POE/Kernel\.pm line 1798\.
:ok 18
:\.\.\.\.\.\.\.\.
:
:This behavior has been verified with perl 5\.8\.0 on Linux\, MacOS X\, and
:FreeBSD\.

I can't reproduce this with POE\-0\.22 under Linux\.

My config is below\, in case it helps\.

Hugo


---
Summary of my perl5 \(revision 5\.0 version 8 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.4\.2\-2\, archname=i686\-linux
&nbsp;    uname='linux crypt\.compulink\.co\.uk 2\.4\.2\-2 \#1 sun apr 8 20&#8203;:41&#8203;:30 edt 2001 i686 unknown '
&nbsp;    config\_args='\-des \-Dprefix=/opt/perl\-5\.8\.0 \-Doptimize=\-g \-O6 \-Dusedevel \-Uversiononly'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-DDEBUGGING \-fno\-strict\-aliasing \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64 \-I/usr/include/gdbm'\,
&nbsp;    optimize='\-g \-O6'\,
&nbsp;    cppflags='\-DDEBUGGING \-fno\-strict\-aliasing \-I/usr/include/gdbm'
&nbsp;    ccversion=''\, gccversion='2\.96 20000731 \(Red Hat Linux 7\.1 2\.96\-81\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-lndbm \-lgdbm \-ldl \-lm \-lc \-lcrypt \-lutil
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lc \-lcrypt \-lutil
&nbsp;    libc=/lib/libc\-2\.2\.2\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.2\.2'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: DEBUGGING USE\_LARGE\_FILES
&nbsp;  Built under linux
&nbsp;  Compiled at Jul 19 2002 15&#8203;:54&#8203;:15
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /opt/perl\-5\.8\.0/lib/5\.8\.0/i686\-linux
&nbsp;    /opt/perl\-5\.8\.0/lib/5\.8\.0
&nbsp;    /opt/perl\-5\.8\.0/lib/site\_perl/5\.8\.0/i686\-linux
&nbsp;    /opt/perl\-5\.8\.0/lib/site\_perl/5\.8\.0
&nbsp;    /opt/perl\-5\.8\.0/lib/site\_perl
&nbsp;    \.

**Comment by p5pRT** [comment]
### From troc@netrus.net

On Tue\, Aug 06\, 2002 at 12&#8203;:21&#8203;:53AM \+0100\, hv@&#8203;crypt\.org wrote&#8203;:
> Rocco Caputo \(via RT\) \<perlbug@&#8203;perl\.org> wrote&#8203;:
> :Tests for the POE 0\.22 distribution \(on the CPAN\) have been working
> :fine with perl 5\.6\.1 for several weeks\.  They pass for perl 5\.8\.0\, but
> :they generate a disturbing warning in the process&#8203;:
> :
> :2\) eyrie&#8203;:\~/perl/poe% /usr/local/perl\-580/bin/perl t/08\_errors\.t
> :1\.\.59
> :ok 1
> :\.\.\.\.\.\.\.
> :ok 17
> :Argument "\,\-\-\-\-\- SELECT BITS IN \-\-\-\-\-\\n" isn't numeric in array \\
> :        element at /home/troc/perl/poe/POE/Kernel\.pm line 1798\.
> :ok 18
> :\.\.\.\.\.\.\.\.
> :
> :This behavior has been verified with perl 5\.8\.0 on Linux\, MacOS X\, and
> :FreeBSD\.
> 
> I can't reproduce this with POE\-0\.22 under Linux\.
> 
> My config is below\, in case it helps\.

\[\.\.\.\]

>     usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef

\[\.\.\.\.\]

It was discovered after I posted the report that the problem only
occurs with threads enabled\.  From my configuration&#8203;:

usethreads=define use5005threads=undef useithreads=define usemultiplicity=define

I also have usereentrant defined\.

\-\- Rocco Caputo / troc@&#8203;pobox\.com / poe\.perl\.org / poe\.sf\.net

**Comment by p5pRT** [comment]
### From arthur@contiller.se


On tisdag\, augusti 6\, 2002\, at 01&#8203;:21 \, hv@&#8203;crypt\.org wrote&#8203;:

> Rocco Caputo \(via RT\) \<perlbug@&#8203;perl\.org> wrote&#8203;:
> :Tests for the POE 0\.22 distribution \(on the CPAN\) have been working
> :fine with perl 5\.6\.1 for several weeks\.  They pass for perl 5\.8\.0\, but
> :they generate a disturbing warning in the process&#8203;:
> :
> :2\) eyrie&#8203;:\~/perl/poe% /usr/local/perl\-580/bin/perl t/08\_errors\.t
> :1\.\.59
> :ok 1
> :\.\.\.\.\.\.\.
> :ok 17
> :Argument "\,\-\-\-\-\- SELECT BITS IN \-\-\-\-\-\\n" isn't numeric in array \\
> :        element at /home/troc/perl/poe/POE/Kernel\.pm line 1798\.
> :ok 18
> :\.\.\.\.\.\.\.\.
> :
> :This behavior has been verified with perl 5\.8\.0 on Linux\, MacOS X\, and
> :FreeBSD\.
>
> I can't reproduce this with POE\-0\.22 under Linux\.
>
> My config is below\, in case it helps\.
>
> Hugo
> \-\-\-

This is ithread specific\, I think the bug is that we for some strange 
reason\, don't move constants to the pad in peep\, meaning a op 
combination stops the peephole optimizer\.

I need to find some time to debug it but haven't yet\.

Arthur

**Comment by p5pRT** [comment]
arthur@contiller.se - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From arthur@contiller.se

Fixed with patch 18820

&nbsp;        Fixes bug \#15654
&nbsp;        What happend was that a constant was freed\, the pad released but 
&nbsp;        the pad slot still held the SV\, when pad slot was reallocated
&nbsp;        to be a target for a stringify\, it did a sv\_setpv on the target
&nbsp;        and the original SV was whiped out\. When this SV was later on
&nbsp;        to new places using the constant\, they got the wrong value\.
&nbsp;        By replacing pad\_free with pad\_swipe for these cases\, we
&nbsp;        won't have such a problem\. \(pad\_swipe also removes the
&nbsp;        pointer to the original SV\)\.

**Comment by p5pRT** [comment]
arthur@contiller.se - Status changed from 'open' to 'resolved'
